### PR TITLE
RSA: support for Key wrapping + PKCS#8

### PR DESCRIPTION
**Pull request by bgrieder**
### Features

- Add support for RSA key generation, key-wrapping, and PKCS#8 import/export

### Fixed

- Fixed export of Curve25519Secret
- Aligned NIST Curves PKC8 import/export with the pkcs8 crate

### CODE REVIEW

File: src/asymmetric_crypto/mod.rs [hunk 1]
```diff
@@ -1,9 +1,24 @@
...
+}
+
+impl<T: PrivateKey> From<&T> for Box<dyn PublicKey>
+where
+    T::PublicKey: 'static,
+{
+    fn from(private_key: &T) -> Self {
+        Box::new(private_key.public_key())
+    }
+}
```
Comment (tbrezot) [inline src/asymmetric_crypto/mod.rs:24]: Wouldn't it be better to implement this for each type instead of using a dynamic dispatch?

File: src/asymmetric_crypto/rsa/private_key.rs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+    /// Get the key length which is the modulus size in bits
+    pub fn key_length(&self) -> RsaKeyLength {
+        match self.0.n().bits() {
+            2048 => RsaKeyLength::Modulus2048,
+            3072 => RsaKeyLength::Modulus3072,
+            4096 => RsaKeyLength::Modulus4096,
+            _ => panic!("Invalid RSA key length; this should never happen"),
+        }
+    }
```
Comment (tbrezot) [inline src/asymmetric_crypto/rsa/private_key.rs:36]: You could also have stored the key inside an `enum`:
```rust
pub enum RsaPrivateKey {
    Modulus2048(rsa::RsaPrivateKey),
    ...
}
```
Is there an issue with this?

File: src/asymmetric_crypto/rsa/private_key.rs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            }
+        })
+    }
+}
+
+impl PrivateKey for RsaPrivateKey {
+    type PublicKey = RsaPublicKey;
+
+    fn public_key(&self) -> Self::PublicKey {
+        self.0.to_public_key().into()
```
Comment (tbrezot) [inline src/asymmetric_crypto/rsa/private_key.rs:78]: If you make the `RsaPublicKey`'s internal `pub(crate)`, you can avoid dynamic dispatch:
```suggestion
        RsaPublicKey(self.0.to_public_key())
```

File: src/asymmetric_crypto/rsa/private_key.rs [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+impl RsaPrivateKey {
+    /// Get the public key
+    ///
+    /// This is a facade for [`PrivateKey::public_key`]
+    pub fn public_key(&self) -> <RsaPrivateKey as crate::asymmetric_crypto::PrivateKey>::PublicKey {
+        PrivateKey::public_key(self)
+    }
+}
+
+impl ZeroizeOnDrop for RsaPrivateKey {}
```
Comment (tbrezot) [inline src/asymmetric_crypto/rsa/private_key.rs:92]: Is the `rsa::RsaPrivateKey` already zeroized on drop?

File: src/asymmetric_crypto/rsa/private_key.rs [hunk 5]
```diff
@@ -0,0 +1,151 @@
+use digest::{Digest, DynDigest};
+use pkcs8::SecretDocument;
+use rand_chacha::rand_core::CryptoRngCore;
+use rand_core::RngCore;
```
Comment (tbrezot) [inline src/asymmetric_crypto/rsa/private_key.rs:4]: use the reexport version instead.

File: src/asymmetric_crypto/rsa/private_key.rs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+impl pkcs8::EncodePrivateKey for RsaPrivateKey {
+    fn to_pkcs8_der(&self) -> pkcs8::Result<SecretDocument> {
+        self.0.to_pkcs8_der()
+    }
+
+    fn to_pkcs8_encrypted_der(
+        &self,
+        rng: impl rand_core::CryptoRng + RngCore,
```
Comment (tbrezot) [inline src/asymmetric_crypto/rsa/private_key.rs:140]: The trait `CryptoRngCore` is made to avoid that. Plus, you shouldn't take ownership of the `rng`.
```suggestion
        rng: &mut impl CryptoRngCore,
```

File: src/asymmetric_crypto/rsa/private_key.rs [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    // Splits the input into two parts. The first is the wrapped AES key, and the
+    // second is the wrapped target key. The length of the first part is equal to
+    // the length of the unwrapping RSA key.
+    let encapsulation_bytes_len = ((rsa_private_key.key_length() as i32) / 8) as usize;
+    // Un-wraps the temporary AES key from the first part with the private RSA key
+    // using CKM_RSA_PKCS_OAEP with parameters of OAEPParams.
+    let aes_key = ckm_rsa_pkcs_oaep::<H>(
+        rsa_private_key,
+        ciphertext[0..encapsulation_bytes_len].as_ref(),
+    )?;
```
Comment (tbrezot) [inline src/asymmetric_crypto/rsa/private_key.rs:123]:
```suggestion
    let aes_key = ckm_rsa_pkcs_oaep::<H>(rsa_private_key, &ciphertext[0..encapsulation_bytes_len])?;
```

File: src/asymmetric_crypto/rsa/private_key.rs [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+///
+/// Also check [https://cloud.google.com/kms/docs/key-wrapping?hl=fr]
+fn ckm_rsa_aes_key_unwrap<H: 'static + Digest + DynDigest + Send + Sync>(
+    rsa_private_key: &RsaPrivateKey,
+    ciphertext: &[u8],
+) -> Result<Zeroizing<Vec<u8>>, CryptoCoreError> {
+    // Splits the input into two parts. The first is the wrapped AES key, and the
+    // second is the wrapped target key. The length of the first part is equal to
+    // the length of the unwrapping RSA key.
+    let encapsulation_bytes_len = ((rsa_private_key.key_length() as i32) / 8) as usize;
```
Comment (tbrezot) [inline src/asymmetric_crypto/rsa/private_key.rs:117]: Strange:
```suggestion
    let encapsulation_bytes_len = rsa_private_key.key_length() as usize >> 3;
```

File: src/asymmetric_crypto/rsa/private_key.rs [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    // Un-wraps the temporary AES key from the first part with the private RSA key
+    // using CKM_RSA_PKCS_OAEP with parameters of OAEPParams.
+    let aes_key = ckm_rsa_pkcs_oaep::<H>(
+        rsa_private_key,
+        ciphertext[0..encapsulation_bytes_len].as_ref(),
+    )?;
+    // Un-wraps the target key from the second part with the temporary AES key using
+    // CKM_AES_KEY_WRAP_KWP ([AES KEYWRAP] section 6.3). Zeroizes the temporary
+    // AES key. Returns the handle to the newly unwrapped target key.
```
Comment (tbrezot) [inline src/asymmetric_crypto/rsa/private_key.rs:126]: Are you really zeroizing the temporary AES key? It should be automatic as it is zeroized on drop. Plus what is the argument not to return a `SymmetricKey<TARGET_LENGTH>`?

File: src/asymmetric_crypto/rsa/public_key.rs [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+            RsaKeyWrappingAlgorithm::Aes256Sha256 => {
+                ckm_rsa_aes_key_wrap::<sha2::Sha256, { 128 / 8 }, R>(rng, self, key_material)?
+            }
+            RsaKeyWrappingAlgorithm::Aes256Sha1 => {
+                ckm_rsa_aes_key_wrap::<sha1::Sha1, { 256 / 8 }, R>(rng, self, key_material)?
+            }
+            RsaKeyWrappingAlgorithm::Aes256Sha3 => {
+                ckm_rsa_aes_key_wrap::<sha3::Sha3_256, { 256 / 8 }, R>(rng, self, key_material)?
+            }
+        })
```
Comment (tbrezot) [inline src/asymmetric_crypto/rsa/public_key.rs:55]: You don't need unwrapping errors:
```suggestion
        match wrapping_algorithm {
            RsaKeyWrappingAlgorithm::Pkcs1v1_5 => self
                .0
                .encrypt(&mut *rng, rsa::Pkcs1v15Encrypt, key_material)
                .map_err(CryptoCoreError::from),
            RsaKeyWrappingAlgorithm::OaepSha256 => {
                ckm_rsa_pkcs_oaep::<sha2::Sha256, R>(rng, self, key_material)
            }
            RsaKeyWrappingAlgorithm::OaepSha1 => {
                ckm_rsa_pkcs_oaep::<sha1::Sha1, R>(rng, self, key_material)
            }
            RsaKeyWrappingAlgorithm::OaepSha3 => {
                ckm_rsa_pkcs_oaep::<sha3::Sha3_256, R>(rng, self, key_material)
            }
            RsaKeyWrappingAlgorithm::Aes256Sha256 => {
                ckm_rsa_aes_key_wrap::<sha2::Sha256, { 128 / 8 }, R>(rng, self, key_material)
            }
            RsaKeyWrappingAlgorithm::Aes256Sha1 => {
                ckm_rsa_aes_key_wrap::<sha1::Sha1, { 256 / 8 }, R>(rng, self, key_material)
            }
            RsaKeyWrappingAlgorithm::Aes256Sha3 => {
                ckm_rsa_aes_key_wrap::<sha3::Sha3_256, { 256 / 8 }, R>(rng, self, key_material)
            }
        }
```

File: src/asymmetric_crypto/rsa/public_key.rs [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    const AES_KEY_LENGTH: usize,
+    R: CryptoRngCore,
+>(
+    rng: &mut R,
+    rsa_public_key: &RsaPublicKey,
+    key_material: &Zeroizing<Vec<u8>>,
+) -> Result<Vec<u8>, CryptoCoreError> {
+    // Generates a temporary random AES key of ulAESKeyBits length.  This key is not
+    // accessible to the user - no handle is returned.
+    let key_encryption_key = SymmetricKey::<AES_KEY_LENGTH>::new(rng);
```
Comment (tbrezot) [inline src/asymmetric_crypto/rsa/public_key.rs:93]:
```suggestion
    let encryption_key = SymmetricKey::<AES_KEY_LENGTH>::new(rng);
```

File: src/asymmetric_crypto/rsa/public_key.rs [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    // Wraps the target key with the temporary AES key using CKM_AES_KEY_WRAP_KWP
+    // ([AES KEYWRAP] section 6.3). PKCS#11 CKM_AES_KEY_WRAP_KWP is identical to
+    // tRFC 5649
+    let mut ciphertext = key_wrap(key_material, &key_encryption_key)?;
+    //Wraps the AES key with the wrapping RSA key using CKM_RSA_PKCS_OAEP with
```
Comment (tbrezot) [inline src/asymmetric_crypto/rsa/public_key.rs:98]: Why not passing the wrapping AES key as argument? Instead of your `key_material: Zeroizing<Vec<u8>>`, passing a `wrapping_aes_key: SymmetricKey<32>`.

File: src/asymmetric_crypto/rsa/public_key.rs [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    // parameters of OAEPParams. Zeroizes the temporary AES key (automatically
+    // done by the conversion into())
+    let mut wrapped_kwk =
+        ckm_rsa_pkcs_oaep::<H, R>(rng, rsa_public_key, &key_encryption_key.into())?;
+    // Concatenates two wrapped keys and outputs the concatenated blob.
+    // The first is the wrapped AES key, and the second is the wrapped target key.
+    wrapped_kwk.append(&mut ciphertext);
```
Comment (tbrezot) [inline src/asymmetric_crypto/rsa/public_key.rs:105]: You may want to allocate the correct size directly.

File: src/error.rs [hunk 14]
```diff
@@ -50,53 +52,40 @@ pub enum CryptoCoreError {
...
-                "when encrypting, plaintext of {plaintext_len} bytes is too big, max is {max} \
-                 bytes"
-            ),
-            #[cfg(any(feature = "certificate", feature = "nist_curves"))]
-            Self::Pkcs8Error(err) => write!(f, "when converting to PKCS8, {err}"),
             #[cfg(any(feature = "certificate", feature = "nist_curves"))]
-            Self::Certificate(err) => write!(f, "when building certificate, {err}"),
-            Self::CiphertextTooSmallError {
+            CryptoCoreError::Certificate(err) => write!(f, "when building certificate, {err}"),
+            CryptoCoreError::CiphertextTooSmallError {
```
Comment (tbrezot) [inline src/error.rs:57]: Why this change?

File: src/asymmetric_crypto/curves/nist/private_key.rs
[the remark this answers is not in the document]
Comment (bgrieder) [reply src/asymmetric_crypto/curves/nist/private_key.rs:1]: Basically lines 36-117 were moved to 208-298 and deprecated; those were partially implementing PKCS#8 support

Support for PKCS#8 was added by implementing EncodedPrivateKey and DecodedPrivateKey lines 168-208

File: src/asymmetric_crypto/mod.rs (hunk 1 above)
Comment (bgrieder) [reply src/asymmetric_crypto/mod.rs:24]: I just left it this for compatibility but I think this should be deprecated. From now on we should use `public_key()` on the private key.
https://github.com/Cosmian/crypto_core/issues/62

File: src/asymmetric_crypto/rsa/private_key.rs (hunk 2 above)
Comment (bgrieder) [reply src/asymmetric_crypto/rsa/private_key.rs]: This makes the high-level RsaPrivateKey object an enum, rather than a struct, which is a completely different approach. I guess this would work too.  The current implementation mirrors the implementation of the rsa crate and the ones we have for symmetric keys where then length is a property

File: src/asymmetric_crypto/rsa/public_key.rs (hunk 11 above)
Comment (bgrieder) [reply src/asymmetric_crypto/rsa/public_key.rs:94]: C'est la terminologies officielle de la RFC 5649

File: src/asymmetric_crypto/rsa/public_key.rs (hunk 12 above)
Comment (bgrieder) [reply src/asymmetric_crypto/rsa/public_key.rs:99]: Our implementation of RFC 5649 does not use the SymmetricKey structure. That should be changed first. then this.
I do not mid opening an issue on this

Comment (bgrieder) [reply src/asymmetric_crypto/rsa/public_key.rs:99]: Updated https://github.com/Cosmian/crypto_core/issues/61

File: src/asymmetric_crypto/rsa/public_key.rs (hunk 13 above)
Comment (bgrieder) [reply src/asymmetric_crypto/rsa/public_key.rs:106]: Not sure how to do this since ckm_rsa_pkcs_oaep allocates the vec, and it would be awkward for this function to have as an argument the ciphertext size

File: src/error.rs (hunk 14 above)
Comment (bgrieder) [reply src/error.rs:57]: Because this is the recommended way of doing it

File: src/asymmetric_crypto/rsa/private_key.rs (hunk 4 above)
Comment (bgrieder) [reply src/asymmetric_crypto/rsa/private_key.rs:94]: Yes

File: src/asymmetric_crypto/rsa/private_key.rs (hunk 8 above)
Comment (bgrieder) [reply src/asymmetric_crypto/rsa/private_key.rs:119]: The enum returns an i32, which represent buts, we want bytes and cast that as an usize

File: src/asymmetric_crypto/rsa/private_key.rs (hunk 9 above)
Comment (bgrieder) [reply src/asymmetric_crypto/rsa/private_key.rs:125]: The AES key is a Zeroizing<Vec>;  our RFC 5649 impl does not use SymmetricKey. See issue #61

File: src/asymmetric_crypto/rsa/public_key.rs (hunk 10 above)
Comment (bgrieder) [reply src/asymmetric_crypto/rsa/public_key.rs:56]: ok

File: src/error.rs (hunk 14 above)
Comment (tbrezot) [reply src/error.rs:57]: I thought this was this opposite: https://rust-lang.github.io/rust-clippy/master/index.html#use_self

Comment (bgrieder) [reply src/error.rs:57]: When you let Rust analyzer generate match cases for Enums, it will ignore the Self ones and generate then with the fully qualified name. I believe rust analyzer uses clippy recommendations behind the hood. 
I honestly believe this is the correct choice for a match where you do not want ambiguity in the enumeration; the opposite would be true for a "constructor" which is what you example shows (even though there is little to none inheritance in Rust)
